pyvista/plotting/composite_mapper.py

```diff
@@ -0,0 +1,769 @@
+                vtk_ref = _vtk.reference(0)  # needed for <=9.0.3
+                block = self.DataObjectFromIndex(index, self._dataset, vtk_ref)
+            else:
+                block = self.DataObjectFromIndex(index, self._dataset)
+        except OverflowError:
+            raise KeyError(f'Invalid block key: {index}') from None
+        if block is None:
+            if index > len(self):
+                raise KeyError(
+                    f'index {index} is out of bounds. There are only {len(self)} blocks.'
+                ) from None
+        return block
+
+    def __getitem__(self, index):
+        """Return a block attribute by its flat index."""
+        return BlockAttributes(self.get_block(index), self)
+
+    def __len__(self):
+        """Return the number of blocks."""
+        return len(self._dataset)
+
+    def __iter__(self):
+        """Return an iterator of all the block attributes."""
+        for ii in range(len(self)):
+            yield self[ii]
+
+
+class CompositePolyDataMapper(_vtk.vtkCompositePolyDataMapper2):
+    """Composite PolyData mapper.
+
+    Parameters
+    ----------
+    dataset : pyvista.MultiBlock
+        Multiblock dataset.
+
+    color_missing_with_nan : bool, optional
+        Color any missing values with the ``nan_color``. This is useful
+        when not all blocks of the composite dataset have the specified
+        ``scalars``.
+
+    interpolate_before_map : bool, optional
+        Enabling makes for a smoother scalars display.  Default is
+        ``True``.  When ``False``, OpenGL will interpolate the
+        mapped colors which can result is showing colors that are
+        not present in the color map.
+
+    """
+
+    def __init__(self, dataset, color_missing_with_nan=None, interpolate_before_map=None):
+        """Initialize this composite mapper."""
+        super().__init__()
+        self.SetInputDataObject(dataset)
+
+        # this must be added to set the color, opacity, and visibility of
+        # individual blocks
+        self._attr = CompositeAttributes(self, dataset)
+        self._dataset = dataset
+
+        if color_missing_with_nan is not None:
+            self.color_missing_with_nan = color_missing_with_nan
+        if interpolate_before_map is not None:
+            self.interpolate_before_map = interpolate_before_map
+
+    @property
+    def interpolate_before_map(self) -> bool:
+        """Return or set the interpolation of scalars before mapping."""
+        return bool(self.GetInterpolateScalarsBeforeMapping())
+
+    @interpolate_before_map.setter
+    def interpolate_before_map(self, value: bool):
+        self.SetInterpolateScalarsBeforeMapping(value)
+
+    @property
+    def block_attr(self) -> CompositeAttributes:
+        """Return the block attributes.
+
+        Examples
+        --------
+        Add a sphere and a cube as a multiblock dataset to a plotter and then
+        change the visibility and color of the blocks.
+
```

Comment:
One-basedness mentioned here too?

Reply:
Mentioned in aee6ffa58ee0f26d183b464f761ef056f110cae2.